pkg/cloudprovider/providers/azure/azure.go

```diff
@@ -0,0 +1,143 @@
+package azure
+
+import (
+	"encoding/json"
+	"io"
+
+	"k8s.io/kubernetes/pkg/cloudprovider"
+
+	"github.com/Azure/azure-sdk-for-go/arm/compute"
+	"github.com/Azure/azure-sdk-for-go/arm/network"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/golang/glog"
+)
+
+const AzureProviderName = "azure"
+
+type AzureConfig struct {
+	Cloud             string `json:"cloud"`
+	TenantID          string `json:"tenantId"`
+	SubscriptionID    string `json:"subscriptionId"`
+	ResourceGroup     string `json:"resourceGroup"`
+	Location          string `json:"location"`
+	VnetName          string `json:"vnetName"`
+	SubnetName        string `json:"subnetName"`
+	SecurityGroupName string `json:"securityGroupName"`
+	RouteTableName    string `json:"routeTableName"`
+
+	AdClientID     string `json:"adClientId"`
+	AdClientSecret string `json:"adClientSecret"`
+	AdTenantID     string `json:"adTenantId"`
+}
+
+type AzureCloud struct {
+	AzureConfig
+	Environment             azure.Environment
+	RoutesClient            network.RoutesClient
+	SubnetsClient           network.SubnetsClient
+	InterfacesClient        network.InterfacesClient
+	RouteTablesClient       network.RouteTablesClient
+	LoadBalancerClient      network.LoadBalancersClient
+	PublicIPAddressesClient network.PublicIPAddressesClient
+	SecurityGroupsClient    network.SecurityGroupsClient
+	VirtualMachinesClient   compute.VirtualMachinesClient
+}
+
+func init() {
+	cloudprovider.RegisterCloudProvider(AzureProviderName, func(configReader io.Reader) (cloudprovider.Interface, error) {
+		var az AzureCloud
+		err := json.NewDecoder(configReader).Decode(&az)
+		if err != nil {
+			glog.Errorf("init: failed to load config")
+			return nil, err
+		}
+
+		switch az.Cloud {
+		case "fairfax":
+			az.Environment = azure.USGovernmentCloud
+		case "mooncake":
```

Comment:
Agreed, let's sort it out here and then I'll consume it: https://github.com/Azure/go-autorest/pull/80